src/database/database-leveldb.cpp

```diff
@@ -98,6 +100,118 @@ void Database_LevelDB::listAllLoadableBlocks(std::vector<v3s16> &dst)
 	delete it;
 }
 
+PlayerDatabaseLevelDB::PlayerDatabaseLevelDB(const std::string &savedir)
+{
+	leveldb::Options options;
+	options.create_if_missing = true;
+	leveldb::Status status = leveldb::DB::Open(options,
+		savedir + DIR_DELIM + "players.db", &m_database);
+	ENSURE_STATUS_OK(status);
+}
+
+PlayerDatabaseLevelDB::~PlayerDatabaseLevelDB()
+{
+	delete m_database;
+}
+
+void PlayerDatabaseLevelDB::savePlayer(RemotePlayer *player)
+{
+	/*
+	u8 version = 1
+	u16 hp
+	v3f position
+	f32 pitch
+	f32 yaw
+	u16 breath
+	u32 attribute_count
+	for each attribute {
+		std::string name
+		std::string (long) value
+	}
+	std::string (long) serialized_inventory
+	*/
+
+	std::ostringstream os;
+	writeU8(os, 1);
+
+	PlayerSAO *sao = player->getPlayerSAO();
+	sanity_check(sao);
+	writeU16(os, sao->getHP());
+	writeV3F32(os, sao->getBasePosition());
+	writeF32(os, sao->getLookPitch());
+	writeF32(os, sao->getRotation().Y);
+	writeU16(os, sao->getBreath());
+
+	StringMap stringvars = sao->getMeta().getStrings();
+	writeU32(os, stringvars.size());
+	for (const auto &it : stringvars) {
+		os << serializeString(it.first);
+		os << serializeLongString(it.second);
+	}
+
+	std::ostringstream serialized_inventory;
+	player->inventory.serialize(serialized_inventory);
+	os << serializeLongString(serialized_inventory.str());
```

Comment:
You can directly serialize the inventory into `os`, no need for the extra string.

Reply:
Changed, I was using a separate string to allow for extra items to be added without bumping the version number.